fix: Test DB Creation when meltano install is ran

visch opened the pull request:
Closes #6383 

- [x] Verify this fails on the version of meltano specified in #6383
- [x] Figure out why the sqllite db isn't being created
- [x] Remove the TODO

visch committed:
Testing DB Creation in place while disabling the default DB URI, not clear why this isn't creating the DB and still working

visch commented on tests/meltano/cli/test_install.py:
```diff
@@ -242,3 +244,12 @@ def test_clean_install(
+def test_new_folder_should_autocreate_on_install(project, cli_runner, un_engine_uri):
+    # We had a case defined https://github.com/meltano/meltano/issues/6383 that caused the meltano db to not be recreated
+    shutil.rmtree(".meltano")
+    result = cli_runner.invoke(cli, ["install", "--clean"])
```
The `.meltano` directory is being recreated in the temp directory, but there's no database being created.

I tested with the config command below `meltano config meltano list` and the URI is set properly (SQL Lite DB in the Temp Directory) not sure what's going on here

visch commented on tests/meltano/cli/test_install.py:
```diff
@@ -242,3 +244,12 @@ def test_clean_install(
+    # We had a case defined https://github.com/meltano/meltano/issues/6383 that caused the meltano db to not be recreated
+    shutil.rmtree(".meltano")
+    result = cli_runner.invoke(cli, ["install", "--clean"])
+    result = cli_runner.invoke(cli, ["config", "meltano", "list"])
```
Need to remember to remove this line

visch commented on tests/meltano/cli/test_install.py:
```diff
@@ -242,3 +244,12 @@ def test_clean_install(
             assert len(mappers) == 3
+
+
+# TODO is install_service the right place for this?
```
Reminder to remove this

visch commented on tests/fixtures/db/__init__.py:
```diff
@@ -18,6 +19,17 @@ def engine_uri_env(engine_uri):
         monkeypatch.undo()
 
 
+# When we want to test functinality that doesn't use the current DB URI
```
Fix spelling `functionality`

visch committed:
Project is now created after Database URI is changed to the default most users will have set during development

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install

visch commented:
> # [Codecov](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano) Report
> > Merging [#6407](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano) ([297435c](https://github.com/meltano/meltano/commit/297435c1e214d827a171c6efb4fc4dba84003d76)) into [main](https://codecov.io/gh/meltano/meltano/commit/455ba3bec860a8b71956f33bf6039c2e42c69552?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano) ([455ba3b](https://github.com/meltano/meltano/commit/455ba3bec860a8b71956f33bf6039c2e42c69552)) will **decrease** coverage by `0.02%`.
> > The diff coverage is `100.00%`.
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main    #6407      +/-   ##
> ==========================================
> - Coverage   89.06%   89.03%   -0.03%     
> ==========================================
>   Files         277      277              
>   Lines       19708    19735      +27     
>   Branches     1944     1944              
> ==========================================
> + Hits        17552    17572      +20     
> - Misses       1831     1835       +4     
> - Partials      325      328       +3     
> ```
> 
> [Impacted Files](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano)	Coverage Δ	
> [tests/fixtures/core.py](https://codecov.io/gh/meltano/meltano/pull/6407/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano#diff-dGVzdHMvZml4dHVyZXMvY29yZS5weQ==)	`90.87% <100.00%> (+0.43%)`	⬆️
> [tests/fixtures/db/__init__.py](https://codecov.io/gh/meltano/meltano/pull/6407/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano#diff-dGVzdHMvZml4dHVyZXMvZGIvX19pbml0X18ucHk=)	`100.00% <100.00%> (ø)`	
> [tests/meltano/cli/test_install.py](https://codecov.io/gh/meltano/meltano/pull/6407/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano#diff-dGVzdHMvbWVsdGFuby9jbGkvdGVzdF9pbnN0YWxsLnB5)	`98.24% <100.00%> (+0.11%)`	⬆️
> [tests/fixtures/docker/snowplow.py](https://codecov.io/gh/meltano/meltano/pull/6407/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano#diff-dGVzdHMvZml4dHVyZXMvZG9ja2VyL3Nub3dwbG93LnB5)	`75.38% <0.00%> (-9.24%)`	⬇️
> [tests/fixtures/cli.py](https://codecov.io/gh/meltano/meltano/pull/6407/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano#diff-dGVzdHMvZml4dHVyZXMvY2xpLnB5)	`97.61% <0.00%> (-2.39%)`	⬇️
> [Continue to review full report at Codecov](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=footer&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano). Last update [455ba3b...297435c](https://codecov.io/gh/meltano/meltano/pull/6407?src=pr&el=lastupdated&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=meltano).

Tests have coverage as well? 👁️

WillDaSilva commented:
@visch merging `meltano:main` into your branch should resolve your coverage issue.

WillDaSilva commented on tests/fixtures/core.py:
```diff
@@ -472,6 +472,30 @@ def project(test_dir, project_init_service):
 
+# Copied the fixture above in order to change the scope
+# Frees us from side affects from other tests in the same class
+@pytest.fixture(scope="function")
```
Instead of copying the body of the function, could you please move the body into a new [`contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) function, then have the fixtures each use it like so:

```python
@pytest.fixture(scope="my scope")
def fixture_name(...):
    with my_context_manager():
        yield
```

visch replied:
Yes, much better! Thank you

WillDaSilva commented on tests/fixtures/db/__init__.py:
```diff
@@ -18,6 +18,17 @@ def engine_uri_env(engine_uri):
 
 
+@pytest.fixture
+def un_engine_uri():
+    """When we want to test functionality that doesn't use the current DB URI"""
+    monkeypatch = MonkeyPatch()
```
```suggestion
def un_engine_uri(monkeypatch):
    """When we want to test functionality that doesn't use the current DB URI"""
```

visch replied:
Refactored the engine_uri function and this one to use the monkeypatch fixture, and use context managers instead! Thanks for the suggestion earlier I use context managers a lot but I haven't written many! I appreciate the code review

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install

visch committed:
Refactor project fixture to get rid of duplciated code

visch committed:
Merge branch '6383-test-db-creation-during-install' of github.com:AutoIDM/meltano into 6383-test-db-creation-during-install

visch committed:
Cleanup some lint rule suggestions

visch committed:
Monkey patch not needed if you're setting a environment variable for the entire session

visch commented on tests/fixtures/db/__init__.py:
```diff
@@ -9,13 +9,17 @@
-        yield
-    finally:
-        monkeypatch.undo()
+def engine_uri_env(engine_uri, monkeypatch) -> None:
```
Gave up on passing the `flake8` ```DAR101 Missing parameter(s) in Docstring: - monkeypatch
    with monkeypatch.context() as m:```

visch commented on tests/fixtures/core.py:
```diff
@@ -473,25 +472,17 @@ def project(test_dir, project_init_service):
 
 
 @pytest.fixture(scope="class")
-def project_files(test_dir, compatible_copy_tree):
```
I shouldn't be touching this function, removing the change for this one now

visch committed:
Remove an edit to a fixture we didn't need to touch. Added yeild to the monkey patch so it'd function as expected

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install

WillDaSilva commented:
@visch This PR isn't a draft anymore, but the description suggests there is more work to be done. Is the description out of date, or does that work fall outside the scope of this PR?

visch commented:
> @visch This PR isn't a draft anymore, but the description suggests there is more work to be done. Is the description out of date, or does that work fall outside the scope of this PR?

Got it I didn't realize the description had to stay up to date, now I know thank you! I'll try to comply as much as I can.

Yes this PR is ready to go!

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install

WillDaSilva reviewed (changes requested):
Other than my comment about `un_engine_uri`, this looks good to go! Let me know if you think it's worth keeping it as a fixture.

tayloramurphy commented:
@visch couple of open comments from @WillDaSilva on this PR. Looks like it's mostly ready to go though?

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install

visch commented on tests/fixtures/db/__init__.py:
```diff
@@ -1,21 +1,26 @@
+@pytest.fixture
+def un_engine_uri(monkeypatch) -> Generator:
+    """When we want to test functionality that doesn't use the current DB URI."""
+    with monkeypatch.context() as m:
+        m.delenv("MELTANO_DATABASE_URI")
```
Wow thanks! Monkey patch is cooler than I thought

visch commented on tests/fixtures/db/__init__.py:
```diff
@@ -1,21 +1,26 @@
+
+
+@pytest.fixture
+def un_engine_uri(monkeypatch) -> Generator:
```
Yeah I understand the thinking my thought was that I'd need to explain why the side affect was happening inside of the test with a comment. This way the fixture is self describing. Both work, I'll go with your way as you're probably right that other tests probably won't need to do this

visch replied:
We have to set the `MELTANO_DATABASE_URI`  before the `project` fixture is called because in `/home/visch/git/meltano/src/meltano/core/db.py(26)` (below) 
```
> /home/visch/git/meltano/src/meltano/core/db.py(26)project_engine()
-> if existing_engine:
(Pdb) print (existing_engine)
(Engine(sqlite:////tmp/pytest-of-visch/pytest-14/fixture_db_sqlite_engine_uri0/pytest_meltano.db), sessionmaker(class_='Session', bind=Engine(sqlite:////tmp/pytest-of-visch/pytest-14/fixture_db_sqlite_engine_uri0/pytest_meltano.db), autoflush=True, autocommit=False, expire_on_commit=True))
(Pdb) print(os.getenv("MELTANO_DATABASE_URI"))
None
```

Caused by `_engines` here
src/meltano/core/db.py
```
>>  6 _engines = {}
    5
    4
>>  3 def project_engine(project, default=False) -> tuple(Engine, sessionmaker):
    2     """Creates and register a SQLAlchemy engine for a Meltano project instance."""
    1
  24      breakpoint()
    1     existing_engine = _engines.get(project)
 ```


Doing this via a fixture makes sense because of that. I tried to make this work without `un_engine_uri` and ended up back to this :O. We could add an issue for this, I think _engines sticks around due to the `project* `fixture

More rambling and details here https://github.com/meltano/meltano/pull/6407#issuecomment-1200527808

wdyt @WillDaSilva

WillDaSilva replied:
Reloading a module is definitely worse - keeping `un_engine_uri` as a fixture makes sense given this context. A comment explaining that it must run before the project is created would be appreciated.

visch replied:
Done!

visch committed:
Remove un_engine_uri fixture as this is a cleaner way. Add back the monkeypatch for meltano_database_uri just in case someone runs pytest in a long running process, although unlikely, it wouldn't be fun to debug

visch committed:
DB Engine URI Readded where we were before this PR as this works best, shouldn't have ever touched this

visch commented:
Test will still fail here, not certain why. Did some debugging and the DB_URI looks to be set properly. Still digging...

visch committed:
Back to un_engine_uri as we have a cached project with DB set that we don't want to play with

visch commented:
> Test will still fail here, not certain why. Did some debugging and the DB_URI looks to be set properly. Still digging...
> 
> Running `meltano install` from inside the test directory (outside of pytest) `/tmp/pytest-of-visch/pytest-7/meltano_root0/project_test_new_folder_should_autocreate_on_install` runs properly. So it smells like a env variable issue, but `meltano config list` says other wise when ran with the `cli_runner` from inside `pytest` (via a breakpoint)
> 
> Found it. After Install is called the "project" still exists which is cached somewhere in our test.
> 
> ```
> > /home/visch/git/meltano/src/meltano/core/db.py(26)project_engine()
> -> if existing_engine:
> (Pdb) print (existing_engine)
> (Engine(sqlite:////tmp/pytest-of-visch/pytest-14/fixture_db_sqlite_engine_uri0/pytest_meltano.db), sessionmaker(class_='Session', bind=Engine(sqlite:////tmp/pytest-of-visch/pytest-14/fixture_db_sqlite_engine_uri0/pytest_meltano.db), autoflush=True, autocommit=False, expire_on_commit=True))
> (Pdb) print(os.getenv("MELTANO_DATABASE_URI"))
> None
> ```
> 
> `src/meltano/core/db.py`
> 
> ```
> >>  6 _engines = {}
>     5
>     4
> >>  3 def project_engine(project, default=False) -> tuple(Engine, sessionmaker):
>     2     """Creates and register a SQLAlchemy engine for a Meltano project instance."""
>     1
>   24      breakpoint()
>     1     existing_engine = _engines.get(project)
> ```
> 
> Didn't hit this before as I had DB_ENGINE_URI unset before the project fixture was called. 👁️

Went back to what we had with un_engine_uri, I think these steps may clear up why. We could add an issue for this, I think `_engines` sticks around due to the `project*` fixture

visch commented on tests/meltano/cli/test_install.py:
```diff
@@ -242,3 +245,21 @@ def test_clean_install(
+# https://github.com/meltano/meltano/pull/6407#issuecomment-1200516464
+# For more details
+def test_new_folder_should_autocreate_on_install(
+    un_engine_uri, project_function, cli_runner
```
Tried using `project` instead of `project_function` no luck. I spent more time than I probably should have on this so I think this working is probably good enough. We could make a follow up issue regarding https://github.com/meltano/meltano/pull/6407#issuecomment-1200516464 

wdyt @WillDaSilva  ?

visch commented:
> @visch couple of open comments from @WillDaSilva on this PR. Looks like it's mostly ready to go though?

Addressed the issues, dove into the why and reverted back to what worked! I think we're good to go not sure if we need followup issues here or not

visch committed:
Explain how to use un_engine_uri and why it exists

visch committed:
Merge branch 'main' into 6383-test-db-creation-during-install
